File: leetcode/python/000211.design-add-and-search-words-data-structure.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False


class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    # Time complexity: O(m), m being length of word
    # Space complexity: O(m)
    def addWord(self, word: str) -> None:
        cur = self.root

        for c in word:
            if not c in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]

        cur.isEnd = True

    # At each level, for a '.' character, we might need to explore up to 26 possibilities.
    # But on average, it's way better.
    # Time complexity: O(26^m).
    # Space complexity: O(m)
    def search(self, word: str) -> bool:
        def dfs(idx, root):
            cur = root

            for i in range(idx, len(word)):
                c = word[i]
                if c == ".":
                    for child in cur.children.values():
                        if dfs(i + 1, child):
                            return True
                    return False
                else:
                    if not c in cur.children:
                        return False
                    cur = cur.children[c]

            return cur.isEnd

        return dfs(0, self.root)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: leetcode/typescript/000211.design-add-and-search-words-data-structure.ts
class TrieNode {
  children: Map<string, TrieNode>;
  isEnd: boolean;

  constructor() {
    this.children = new Map();
    this.isEnd = false;
  }
}

class WordDictionary {
  root: TrieNode;

  constructor() {
    this.root = new TrieNode();
  }

  // Time complexity: O(n) -- n = word length
  // Space complexity: O(n)
  addWord(word: string): void {
    let cur = this.root;
    for (const c of word) {
      if (!cur.children.has(c)) {
        cur.children.set(c, new TrieNode());
      }
      cur = cur.children.get(c)!;
    }
    cur.isEnd = true;
  }

  // Time complexity: O(26^m) -- m = word length. Worst case is when all chars of the word are "."
  // Space complexity: O(m)
  search(word: string): boolean {
    function dfs(idx: number, root: TrieNode): boolean {
      let cur = root;
      for (let i = idx; i < word.length; i++) {
        const c = word[i];
        if (c == ".") {
          for (const child of cur.children.values()) {
            if (dfs(i + 1, child)) return true;
          }
          return false;
        }
        if (!cur.children.has(c)) return false;
        cur = cur.children.get(c)!;
      }
      return cur.isEnd;
    }

    return dfs(0, this.root);
  }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */
